File: swift-client/Sources/ParRtDbClient/Mutation.swift
import Foundation

/// Protocol contract constants for mutations. `maxSteps` mirrors
/// server/src/txn.rs `MAX_STEPS` and rust-client/src/in_memory/mod.rs
/// `MAX_STEPS`; the wire-corpus `protocol_constants.max_steps` assertion pins
/// this number — change it only with the server and every client together.
public enum MutationLimits {
    /// Maximum steps in one transaction, counted recursively (scheduled txns
    /// included). The server rejects a larger txn; builders should refuse to
    /// construct one rather than send a doomed request.
    public static let maxSteps = 1024
}

/// Mirrors server/src/dsl.rs::Transaction (rust-client/src/mutation.rs::
/// Transaction) — an ordered list of steps applied atomically by the server's
/// committer. `steps` is required; unknown fields are tolerated (neither the
/// server nor the rust-client Transaction carries `deny_unknown_fields`).
/// The max-steps cap is NOT enforced here — `MutationLimits` (Task 6) and the
/// builder layer own it.
public struct Transaction: Equatable, Codable, Sendable {
    /// The steps, applied in order; any failure rolls the whole txn back.
    public var steps: [Step]

    public init(steps: [Step]) {
        self.steps = steps
    }

    enum CodingKeys: String, CodingKey, CaseIterable {
        case steps
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        steps = try container.decode([Step].self, forKey: .steps)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(steps, forKey: .steps)
    }
}

/// Mirrors server/src/dsl.rs::Step (rust-client/src/mutation.rs::Step) — one
/// write/control step, internally tagged on `"op"`, camelCase tags and fields,
/// unknown fields rejected per variant. Document bodies are
/// `[String: JSONValue]` (serde `Map<String, Value>`).
public enum Step: Equatable, Codable, Sendable {
    /// Insert a new document; result is its id.
    case insert(table: String, doc: [String: JSONValue])
    /// Merge `fields` into an existing document; result is null.
    case patch(table: String, id: String, fields: [String: JSONValue])
    /// Overwrite the whole document; result is null.
    case replace(table: String, id: String, doc: [String: JSONValue])
    /// Delete a document; result is null.
    case delete(table: String, id: String)
    /// Precondition: the row must be at exactly `version`.
    case expectVersion(table: String, id: String, version: Int64)
    /// Precondition: no row may match the index eq-prefix.
    case expectAbsent(table: String, index: String, eq: [JSONValue])
    /// Insert-or-patch keyed by an index eq-prefix match.
    case upsert(
        table: String, index: String, eq: [JSONValue],
        insert: [String: JSONValue], patch: [String: JSONValue]
    )
    /// Patch every row matching `filter` (at most `limit`, default server cap).
    case patchByQuery(
        table: String, filter: FilterExpr, patch: [String: JSONValue], limit: UInt32?
    )
    /// Delete every row matching `filter` (same `limit` semantics).
    case deleteByQuery(table: String, filter: FilterExpr, limit: UInt32?)
    /// Schedule `txn` to run later.
    case schedule(when: ScheduleWhen, txn: Transaction)
    /// Cancel a previously scheduled job.
    case cancelSchedule(id: String)
    /// Start a durable workflow run.
    case startWorkflow(spec: WorkflowSpec)
    /// Cancel a workflow run.
    case cancelWorkflow(id: String)
    /// Restore a soft-deleted row (only legal on a `softDelete` table).
    case undelete(table: String, id: String)

    enum CodingKeys: String, CodingKey, CaseIterable {
        case op, table, doc, id, fields, version, index, eq, insert, patch
        case filter, limit, when, txn, spec
    }

    // swiftlint:disable:next cyclomatic_complexity function_body_length
    public init(from decoder: Decoder) throws {
        let payload = try taggedEnumPayload("Step", tagKey: "op", from: decoder)
        let container = try decoder.container(keyedBy: CodingKeys.self)
        switch payload.tag {
        case "insert":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "doc"]
            )
            self = try .insert(
                table: container.decode(String.self, forKey: .table),
                doc: container.decode([String: JSONValue].self, forKey: .doc)
            )
        case "patch":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "id", "fields"]
            )
            self = try .patch(
                table: container.decode(String.self, forKey: .table),
                id: container.decode(String.self, forKey: .id),
                fields: container.decode([String: JSONValue].self, forKey: .fields)
            )
        case "replace":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "id", "doc"]
            )
            self = try .replace(
                table: container.decode(String.self, forKey: .table),
                id: container.decode(String.self, forKey: .id),
                doc: container.decode([String: JSONValue].self, forKey: .doc)
            )
        case "delete":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "id"]
            )
            self = try .delete(
                table: container.decode(String.self, forKey: .table),
                id: container.decode(String.self, forKey: .id)
            )
        case "expectVersion":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "id", "version"]
            )
            self = try .expectVersion(
                table: container.decode(String.self, forKey: .table),
                id: container.decode(String.self, forKey: .id),
                version: container.decode(Int64.self, forKey: .version)
            )
        case "expectAbsent":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "index", "eq"]
            )
            self = try .expectAbsent(
                table: container.decode(String.self, forKey: .table),
                index: container.decode(String.self, forKey: .index),
                eq: container.decode([JSONValue].self, forKey: .eq)
            )
        case "upsert":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "index", "eq", "insert", "patch"]
            )
            self = try .upsert(
                table: container.decode(String.self, forKey: .table),
                index: container.decode(String.self, forKey: .index),
                eq: container.decode([JSONValue].self, forKey: .eq),
                insert: container.decode([String: JSONValue].self, forKey: .insert),
                patch: container.decode([String: JSONValue].self, forKey: .patch)
            )
        case "patchByQuery":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "filter", "patch", "limit"]
            )
            self = try .patchByQuery(
                table: container.decode(String.self, forKey: .table),
                filter: container.decode(FilterExpr.self, forKey: .filter),
                patch: container.decode([String: JSONValue].self, forKey: .patch),
                limit: container.decodeIfPresent(UInt32.self, forKey: .limit)
            )
        case "deleteByQuery":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "filter", "limit"]
            )
            self = try .deleteByQuery(
                table: container.decode(String.self, forKey: .table),
                filter: container.decode(FilterExpr.self, forKey: .filter),
                limit: container.decodeIfPresent(UInt32.self, forKey: .limit)
            )
        case "schedule":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "when", "txn"]
            )
            self = try .schedule(
                when: container.decode(ScheduleWhen.self, forKey: .when),
                txn: container.decode(Transaction.self, forKey: .txn)
            )
        case "cancelSchedule":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "id"]
            )
            self = try .cancelSchedule(id: container.decode(String.self, forKey: .id))
        case "startWorkflow":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "spec"]
            )
            self = try .startWorkflow(spec: container.decode(WorkflowSpec.self, forKey: .spec))
        case "cancelWorkflow":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "id"]
            )
            self = try .cancelWorkflow(id: container.decode(String.self, forKey: .id))
        case "undelete":
            try rejectUnknownVariantFields(
                "Step", variant: payload.tag, keys: payload.keys,
                allowed: ["op", "table", "id"]
            )
            self = try .undelete(
                table: container.decode(String.self, forKey: .table),
                id: container.decode(String.self, forKey: .id)
            )
        case let unknown:
            throw DecodingError.dataCorrupted(
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "Step: unknown op '\(unknown)'"
                )
            )
        }
    }

    // swiftlint:disable:next cyclomatic_complexity function_body_length
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        switch self {
        case let .insert(table, doc):
            try container.encode("insert", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(doc, forKey: .doc)
        case let .patch(table, id, fields):
            try container.encode("patch", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(id, forKey: .id)
            try container.encode(fields, forKey: .fields)
        case let .replace(table, id, doc):
            try container.encode("replace", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(id, forKey: .id)
            try container.encode(doc, forKey: .doc)
        case let .delete(table, id):
            try container.encode("delete", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(id, forKey: .id)
        case let .expectVersion(table, id, version):
            try container.encode("expectVersion", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(id, forKey: .id)
            try container.encode(version, forKey: .version)
        case let .expectAbsent(table, index, eq):
            try container.encode("expectAbsent", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(index, forKey: .index)
            try container.encode(eq, forKey: .eq)
        case let .upsert(table, index, eq, insert, patch):
            try container.encode("upsert", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(index, forKey: .index)
            try container.encode(eq, forKey: .eq)
            try container.encode(insert, forKey: .insert)
            try container.encode(patch, forKey: .patch)
        case let .patchByQuery(table, filter, patch, limit):
            try container.encode("patchByQuery", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(filter, forKey: .filter)
            try container.encode(patch, forKey: .patch)
            try container.encodeIfPresent(limit, forKey: .limit)
        case let .deleteByQuery(table, filter, limit):
            try container.encode("deleteByQuery", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(filter, forKey: .filter)
            try container.encodeIfPresent(limit, forKey: .limit)
        case let .schedule(when, txn):
            try container.encode("schedule", forKey: .op)
            try container.encode(when, forKey: .when)
            try container.encode(txn, forKey: .txn)
        case let .cancelSchedule(id):
            try container.encode("cancelSchedule", forKey: .op)
            try container.encode(id, forKey: .id)
        case let .startWorkflow(spec):
            try container.encode("startWorkflow", forKey: .op)
            try container.encode(spec, forKey: .spec)
        case let .cancelWorkflow(id):
            try container.encode("cancelWorkflow", forKey: .op)
            try container.encode(id, forKey: .id)
        case let .undelete(table, id):
            try container.encode("undelete", forKey: .op)
            try container.encode(table, forKey: .table)
            try container.encode(id, forKey: .id)
        }
    }
}

/// Mirrors rust-client/src/mutation.rs::StepResult — UNTAGGED: one entry of
/// `mutateOk.results`, decoded by shape. Variant order is load-bearing:
/// `upsert` must precede `insert` because a variant matches when all ITS
/// fields decode (extra keys ignored, serde untagged parity) — with `insert`
/// first, `{"id","inserted"}` would be greedily captured by `insert`,
/// silently dropping `inserted`. `null` is the result of
/// patch/delete/expect*/undelete steps.
public enum StepResult: Equatable, Codable, Sendable {
    /// `{"id", "inserted"}` from an upsert step.
    case upsert(id: String, inserted: Bool)
    /// `{"id"}` from an insert step.
    case insert(id: String)
    /// `{"patched", "truncated"}` from a patchByQuery step.
    case patchByQuery(patched: UInt32, truncated: Bool)
    /// `{"deleted", "truncated"}` from a deleteByQuery step.
    case deleteByQuery(deleted: UInt32, truncated: Bool)
    /// `{"scheduleId"}` from a schedule step.
    case schedule(scheduleId: String)
    /// `{"cancelled"}` from a cancelSchedule/cancelWorkflow step.
    case cancelled(cancelled: Bool)
    /// `{"workflowId"}` from a startWorkflow step.
    case workflowId(workflowId: String)
    /// `null` — patch/delete/expect*/undelete results.
    case null

    public init(from decoder: Decoder) throws {
        let value = try JSONValue(from: decoder)
        guard case let .object(fields) = value else {
            if case .null = value {
                self = .null
                return
            }
            throw DecodingError.dataCorrupted(
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "StepResult: no variant matches \(value)"
                )
            )
        }
        // Each shape is tried in rust's declaration order; first match wins.
        // A variant fails when a required field is missing or wrong-typed,
        // exactly like serde's untagged fall-through.
        if let id = fields["id"]?.stringValue, case let .bool(inserted) = fields["inserted"] {
            self = .upsert(id: id, inserted: inserted)
        } else if let id = fields["id"]?.stringValue {
            self = .insert(id: id)
        } else if let patched = fields["patched"]?.uint32Value, case let .bool(truncated) = fields["truncated"] {
            self = .patchByQuery(patched: patched, truncated: truncated)
        } else if let deleted = fields["deleted"]?.uint32Value, case let .bool(truncated) = fields["truncated"] {
            self = .deleteByQuery(deleted: deleted, truncated: truncated)
        } else if let scheduleId = fields["scheduleId"]?.stringValue {
            self = .schedule(scheduleId: scheduleId)
        } else if case let .bool(cancelled) = fields["cancelled"] {
            self = .cancelled(cancelled: cancelled)
        } else if let workflowId = fields["workflowId"]?.stringValue {
            self = .workflowId(workflowId: workflowId)
        } else {
            throw DecodingError.dataCorrupted(
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "StepResult: no variant matches \(value)"
                )
            )
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case let .upsert(id, inserted):
            try container.encode(JSONValue.object(["id": .string(id), "inserted": .bool(inserted)]))
        case let .insert(id):
            try container.encode(JSONValue.object(["id": .string(id)]))
        case let .patchByQuery(patched, truncated):
            try container.encode(
                JSONValue.object([
                    "patched": .int(Int64(patched)), "truncated": .bool(truncated)
                ])
            )
        case let .deleteByQuery(deleted, truncated):
            try container.encode(
                JSONValue.object([
                    "deleted": .int(Int64(deleted)), "truncated": .bool(truncated)
                ])
            )
        case let .schedule(scheduleId):
            try container.encode(JSONValue.object(["scheduleId": .string(scheduleId)]))
        case let .cancelled(cancelled):
            try container.encode(JSONValue.object(["cancelled": .bool(cancelled)]))
        case let .workflowId(workflowId):
            try container.encode(JSONValue.object(["workflowId": .string(workflowId)]))
        case .null:
            try container.encodeNil()
        }
    }
}

extension JSONValue {
    /// UInt32 view of an integral value, nil when negative/overflow/non-integral
    /// (serde `u32` rejects those on the wire). Used by StepResult shape checks.
    var uint32Value: UInt32? {
        if case let .int(int) = self {
            return UInt32(exactly: int)
        }
        return nil
    }
}

File: core/src/mutation.rs
//! Mutation/transaction DSL wire types shared by the server and Rust client
//! (ARC-004 follow-up): `Transaction`, `Step`, `ScheduleWhen`, `ScheduleKind`,
//! `StepRetry`, `AwaitSignalSpec`, `WorkflowStepSpec`, `WorkflowSpec`, and
//! `WorkflowStatus`. These were two hand-kept copies — one in
//! `server/src/dsl.rs`/`server/src/protocol.rs`, one in
//! `rust-client/src/mutation.rs`/`rust-client/src/wire.rs` — verified
//! structurally identical (same variants, same fields, same serde tags)
//! before being moved here. Wire shapes use
//! `#[serde(tag = "op"/"type", rename_all = "camelCase", deny_unknown_fields)]`
//! and are mirrored field-for-field by the ts/python/swift client SDKs.
//!
//! `Step`/`Transaction` carry a `pub fn table()` on the server
//! (`server/src/dsl.rs`) that is NOT ported here as an inherent `impl` —
//! `Step` is a foreign type from each crate's point of view once re-exported,
//! so the orphan rule forces that helper into a per-crate extension trait at
//! its historical call site (see `server/src/dsl.rs::StepTableExt` and the
//! crate-private table helper in `rust-client/src/optimistic.rs`, which is a
//! genuinely different, ExpectAbsent-masking helper and was never a
//! duplicate).
//!
//! `StepResult` (the untagged step-result decode enum) and the chainable
//! `Mutation` builder are rust-client-only ergonomic sugar with no server
//! equivalent (the server only ever produces the raw JSON these decode) and
//! stay in `rust-client/src/mutation.rs`.

use crate::wire::FilterExpr;

/// One write/control step (tagged `{"op": "..."}`, camelCase).
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "op", rename_all = "camelCase", deny_unknown_fields)]
pub enum Step {
    /// Insert a new document; result is its id.
    Insert {
        /// Target table.
        table: String,
        /// The document body.
        doc: serde_json::Map<String, serde_json::Value>,
    },
    /// Merge `fields` into an existing document; result is `null`.
    Patch {
        /// Target table.
        table: String,
        /// Document id.
        id: String,
        /// Keys to merge in.
        fields: serde_json::Map<String, serde_json::Value>,
    },
    /// Overwrite the whole document; result is `null`.
    Replace {
        /// Target table.
        table: String,
        /// Document id.
        id: String,
        /// The full replacement body.
        doc: serde_json::Map<String, serde_json::Value>,
    },
    /// Delete a document; result is `null`.
    Delete {
        /// Target table.
        table: String,
        /// Document id.
        id: String,
    },
    /// Precondition: the row must be at exactly `version`.
    ExpectVersion {
        /// Target table.
        table: String,
        /// Document id.
        id: String,
        /// The required current version.
        version: i64,
    },
    /// Precondition: no row may match the index eq-prefix.
    ExpectAbsent {
        /// Target table.
        table: String,
        /// Index to probe.
        index: String,
        /// The eq-prefix values.
        eq: Vec<serde_json::Value>,
    },
    /// Insert-or-patch keyed by an index eq-prefix match.
    Upsert {
        /// Target table.
        table: String,
        /// Index whose eq-prefix locates the row.
        index: String,
        /// The eq-prefix values.
        eq: Vec<serde_json::Value>,
        /// Body applied when inserting.
        insert: serde_json::Map<String, serde_json::Value>,
        /// Keys merged when the row exists.
        patch: serde_json::Map<String, serde_json::Value>,
    },
    /// Find every row in `table` matching `filter` (the same `FilterExpr` the
    /// read path accepts) and apply `patch` to it, atomically, inside the
    /// serialized committer turn. Visibility matches the read path exactly: an
    /// interactive caller patches only rows they own/collaborate on and that
    /// satisfy the table's `authorize` predicate; a bypass principal (machine
    /// token/admin/scheduled) touches all matching rows. At most `limit` rows
    /// (default server cap 1000); a larger match set patches `limit` and
    /// reports `truncated: true`. Each patched row records a `DocOp`/`WriteSet`
    /// entry, so subscriptions, the op-feed, audit log, and webhooks all fire
    /// per row — the same contract as a per-id `Patch`.
    PatchByQuery {
        /// Target table.
        table: String,
        /// Which rows match.
        filter: FilterExpr,
        /// Keys to merge into every match.
        patch: serde_json::Map<String, serde_json::Value>,
        /// Row cap (server default/max 1000); `None` = server default.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        limit: Option<u32>,
    },
    /// Find every row in `table` matching `filter` and delete it (same
    /// visibility/`limit`/`truncated` semantics as `PatchByQuery`). Enables
    /// server-side cascades and bulk cleanup (e.g. a scheduled job deleting
    /// expired rows by predicate) without a client-side read-all-then-delete.
    DeleteByQuery {
        /// Target table.
        table: String,
        /// Which rows match.
        filter: FilterExpr,
        /// Row cap (server default/max 1000); `None` = server default.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        limit: Option<u32>,
    },
    /// Schedule `txn` to run later (FM-28). The `scheduled_txns` row is
    /// inserted on the OPEN sqlx transaction, so the enqueue commits (or
    /// rolls back) atomically with this txn's document writes. Step result is
    /// `{"scheduleId": "<id>"}`; the job fires through the unchanged
    /// scheduler → `RunScheduled` committer path as the system (bypass)
    /// principal. Nested steps are table-scope-checked recursively at enqueue
    /// (`authorize_txn_tables`) and fully re-validated by `execute_txn` at
    /// fire time.
    Schedule {
        /// One-shot delay/absolute time, a cron expression, or a fixed
        /// interval.
        when: ScheduleWhen,
        /// The nested transaction to fire when due.
        txn: Box<Transaction>,
    },
    /// Cancel a previously scheduled job by id, on the open sqlx transaction.
    /// Step result `{"cancelled": <bool>}` — `false` (not an error) when the
    /// id is missing, already fired, or already cancelled. A fire currently
    /// in flight completes; the job never fires again (the cron finalize
    /// update touches 0 rows).
    CancelSchedule {
        /// The schedule id to cancel.
        id: String,
    },
    /// Start a durable workflow run (FM-29). The `workflows` row is inserted on
    /// the OPEN sqlx transaction — "write doc + start drip" is atomic; a
    /// rolled-back txn leaves no orphan run. Step result `{"workflowId": "<id>"}`.
    /// The spec is validated and table-scope-checked recursively at submit time;
    /// steps fire later as the system (bypass) principal in the committer's
    /// `RunWorkflowAdvance` turn.
    StartWorkflow {
        /// The run's spec, snapshotted per run server-side.
        spec: Box<WorkflowSpec>,
    },
    /// Cancel a workflow run by id, on the open sqlx transaction. Step result
    /// `{"cancelled": <bool>}` — `false` when missing or already terminal. A run
    /// whose advance is in flight stops at its next step boundary.
    CancelWorkflow {
        /// The workflow run id.
        id: String,
    },
    /// Restore a soft-deleted row (FM-33): `UPDATE … SET deleted_at = NULL,
    /// version = version + 1`. `NotFound` when the row is absent; idempotent
    /// `Ok` when it is present and already live. Only legal on a table that
    /// declares `softDelete`. Patch-shaped `DocOp` — the doc re-appears, so
    /// content-bearing subscriptions re-run.
    Undelete {
        /// Target table (must declare `softDelete`).
        table: String,
        /// Document id.
        id: String,
    },
}

/// An ordered list of steps executed atomically by the server's committer.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct Transaction {
    /// The steps, applied in order; any failure rolls the whole txn back.
    pub steps: Vec<Step>,
}

/// How a caller wants a transaction scheduled. Mirrored byte-for-byte in
/// `ts-client/src/protocol.ts`.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(tag = "type", rename_all = "camelCase", deny_unknown_fields)]
pub enum ScheduleWhen {
    /// Fire `ms` milliseconds from now.
    AfterMs {
        /// Milliseconds from now.
        ms: i64,
    },
    /// Fire at this UTC epoch-ms instant (in the past = fire immediately).
    RunAt {
        /// Absolute UTC epoch-ms instant.
        ms: i64,
    },
    /// Fire on this 5-field cron schedule (UTC, min-first). The server validates
    /// the expression; the client does no cron parsing.
    Cron {
        /// 5-field cron expression (UTC, min-first).
        expr: String,
    },
    /// Fire every `every_ms` milliseconds, starting one interval from now.
    /// Missed windows (downtime, pause) are skipped, never backfilled —
    /// each fire re-arms from its actual fire time, like cron recompute.
    Interval {
        /// The fixed recurrence in milliseconds.
        #[serde(rename = "everyMs")]
        every_ms: i64,
    },
}

/// Whether a scheduled job fires once (`ScheduleWhen::AfterMs`/`RunAt`) or
/// repeats (`Cron` on an expression, `Interval` every N ms). Closed domain —
/// was a free `String` (ARC-004/QA-008). Serializes as `"oneshot"` / `"cron"`
/// / `"interval"`, byte-identical to the prior stringly-typed bytes.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ScheduleKind {
    /// Fires once.
    Oneshot,
    /// Repeats on a cron schedule.
    Cron,
    /// Repeats on a fixed interval.
    Interval,
}

impl ScheduleKind {
    /// The wire string (`"oneshot"` / `"cron"` / `"interval"`).
    pub fn as_wire_str(&self) -> &'static str {
        match self {
            ScheduleKind::Oneshot => "oneshot",
            ScheduleKind::Cron => "cron",
            ScheduleKind::Interval => "interval",
        }
    }
}

impl From<ScheduleKind> for &'static str {
    fn from(k: ScheduleKind) -> &'static str {
        k.as_wire_str()
    }
}

impl std::str::FromStr for ScheduleKind {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "oneshot" => Ok(ScheduleKind::Oneshot),
            "cron" => Ok(ScheduleKind::Cron),
            "interval" => Ok(ScheduleKind::Interval),
            other => Err(format!("unknown ScheduleKind: {other}")),
        }
    }
}

/// Per-step retry policy (FM-29). `maxAttempts` counts TOTAL attempts — the
/// first try included. Defaults when a step omits `retry`: 3 attempts, 1s
/// initial backoff doubling to a 60s cap.
#[derive(Debug, Clone, Copy, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct StepRetry {
    /// TOTAL attempts including the first.
    pub max_attempts: u32,
    /// First backoff delay; doubles per retry.
    #[serde(default = "default_initial_retry_ms")]
    pub initial_retry_ms: u64,
    /// Backoff cap.
    #[serde(default = "default_max_retry_ms")]
    pub max_retry_ms: u64,
}

fn default_initial_retry_ms() -> u64 {
    1_000
}

fn default_max_retry_ms() -> u64 {
    60_000
}

impl Default for StepRetry {
    fn default() -> Self {
        Self {
            max_attempts: 3,
            initial_retry_ms: 1_000,
            max_retry_ms: 60_000,
        }
    }
}

/// An `awaitSignal` step's wait declaration (spec §Wire): park the run
/// until a signal named `name` is delivered; `timeoutMs` bounds each wait
/// attempt (omitted = wait indefinitely, cancel is the escape).
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AwaitSignalSpec {
    /// The signal name to wait for.
    pub name: String,
    /// Per-attempt wait bound; `None` = wait indefinitely.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub timeout_ms: Option<u64>,
}

/// One workflow step: either an ordinary `Transaction` or an
/// [`AwaitSignalSpec`] wait (exactly one — `validate_spec` enforces it),
/// plus policy. The txn may itself carry `Schedule`/`CancelSchedule` steps
/// (FM-28 rules apply).
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WorkflowStepSpec {
    /// The step's transaction (`None` on an `awaitSignal` step).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub txn: Option<Transaction>,
    /// The wait declaration (`None` on a `txn` step).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub await_signal: Option<AwaitSignalSpec>,
    /// Per-step policy (server default 3/1s/60s when `None`).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub retry: Option<StepRetry>,
    /// Gate the step until this many ms have passed.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sleep_before_ms: Option<u64>,
}

/// A submitted workflow definition. Stored verbatim per run — a run
/// snapshots its spec, so template edits never drift a live run.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WorkflowSpec {
    /// Run name (operator-facing).
    pub name: String,
    /// Ordered steps; snapshot per run.
    pub steps: Vec<WorkflowStepSpec>,
}

/// Run lifecycle (FM-29). Closed domain (ARC-004/QA-008 pattern — was never a
/// free string). Snake-case wire: pending|running|waiting|success|failed|cancelled.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkflowStatus {
    /// Not yet advanced.
    Pending,
    /// Mid-run (or crashed mid-run, pre-recovery).
    Running,
    /// Parked on an `awaitSignal` step (non-terminal — a matching signal
    /// or cancel resumes/ends it).
    Waiting,
    /// All steps completed.
    Success,
    /// A step exhausted its retries (terminal).
    Failed,
    /// Cancelled by request (terminal).
    Cancelled,
}

impl WorkflowStatus {
    /// The wire string (`"pending"` etc.).
    pub fn as_wire_str(&self) -> &'static str {
        match self {
            WorkflowStatus::Pending => "pending",
            WorkflowStatus::Running => "running",
            WorkflowStatus::Waiting => "waiting",
            WorkflowStatus::Success => "success",
            WorkflowStatus::Failed => "failed",
            WorkflowStatus::Cancelled => "cancelled",
        }
    }
}

impl std::str::FromStr for WorkflowStatus {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "pending" => Ok(WorkflowStatus::Pending),
            "running" => Ok(WorkflowStatus::Running),
            "waiting" => Ok(WorkflowStatus::Waiting),
            "success" => Ok(WorkflowStatus::Success),
            "failed" => Ok(WorkflowStatus::Failed),
            "cancelled" => Ok(WorkflowStatus::Cancelled),
            other => Err(format!("unknown WorkflowStatus: {other}")),
        }
    }
}
